src/base/materialized.ts

```diff
@@ -230,60 +236,200 @@ export class MaterializedChatMessage implements IMaterializedNode {
 		return tokenizer.countMessageTokens({ ...this.toChatMessage(), content: '' });
 	});
 
-	private readonly _text = once(() => {
-		let result = '';
+	private readonly _text = once((): (string | MaterializedChatMesageImage)[] => {
+		let result: (string | MaterializedChatMesageImage)[] = [''];
 		for (const { text, isTextSibling } of textChunks(this)) {
+			if (text instanceof MaterializedChatMesageImage) {
+				result.push(text);
+				// check children, throw for now
+				for (const child of text.children) {
+					if (child instanceof MaterializedChatMessageTextChunk) {
+						result.push(child.text);
+					}
+				}
+				break;
+			}
 			if (
 				text.lineBreakBefore === LineBreakBefore.Always ||
 				(text.lineBreakBefore === LineBreakBefore.IfNotTextSibling && !isTextSibling)
 			) {
-				if (result.length && !result.endsWith('\n')) {
-					result += '\n';
+				let res = result[result.length - 1]
+				if (!(res instanceof MaterializedChatMesageImage) && result.length && !res.endsWith('\n')) {
+					result.push('\n');
 				}
 			}
 
-			result += text.text;
+			if (typeof result[result.length - 1] === 'string') {
+				result[result.length - 1] += text.text;
+			} else {
+				result.push(text.text);
+			}
 		}
 
-		return result.trim();
+		return result;
 	});
 
 	public toChatMessage(): ChatMessage {
+		const content = this.text
+			.filter(element => typeof element === 'string')
+			.join('').trim();
+
+		const images = this._text().filter(element => element instanceof MaterializedChatMesageImage) as MaterializedChatMesageImage[];
+
+		if (images.length > 0) {
+			return {
+				role: ChatRole.User,
+				content: [
+					{ type: 'text', text: content },
+					{
+						type: 'image_url',
+						image_url: { url: images[0].imageUrl, detail: 'high' },
+					}],
+			};
+		}
+
 		if (this.role === ChatRole.System) {
 			return {
 				role: this.role,
-				content: this.text,
+				content,
 				...(this.name ? { name: this.name } : {}),
 			};
 		} else if (this.role === ChatRole.Assistant) {
 			return {
 				role: this.role,
-				content: this.text,
+				content,
 				...(this.toolCalls ? { tool_calls: this.toolCalls } : {}),
 				...(this.name ? { name: this.name } : {}),
 			};
 		} else if (this.role === ChatRole.User) {
 			return {
 				role: this.role,
-				content: this.text,
+				content,
 				...(this.name ? { name: this.name } : {}),
 			};
 		} else if (this.role === ChatRole.Tool) {
 			return {
 				role: this.role,
-				content: this.text,
+				content,
 				tool_call_id: this.toolCallId,
 			};
 		} else {
 			return {
 				role: this.role,
-				content: this.text,
+				content,
 				name: this.name!,
 			};
 		}
 	}
 }
 
+export class MaterializedChatMesageImage implements IMaterializedNode {
+	constructor(
+		public readonly id: number,
+		public readonly role: ChatRole,
+		public readonly imageUrl: string,
+		public readonly width: number,
```

Comment:
since we're not tokenizing here, we actually prolly don't need width and height anymore. we can read the contents of the chat message and get our width/heigh from the image_url in the extension where we tokenize.